File: webapp/src/services/projects.ts
import api from './api';

// 类型定义
export interface Project {
  id: number;
  name: string;
  description?: string;
  config?: Record<string, any>;
  status: string;
  created_at?: string;
  updated_at?: string;
}

export interface ProjectList {
  total: number;
  items: Project[];
}

export interface CreateProjectData {
  name: string;
  description?: string;
  config?: Record<string, any>;
}

export interface UpdateProjectData {
  name?: string;
  description?: string;
  config?: Record<string, any>;
  status?: string;
}

// 项目API服务
export const projectService = {
  // 获取项目列表
  list: async (skip = 0, limit = 100, status?: string): Promise<ProjectList> => {
    const params: Record<string, any> = { skip, limit };
    if (status) params.status = status;
    return api.get('/projects', { params });
  },

  // 获取单个项目
  get: async (id: number): Promise<Project> => {
    return api.get(`/projects/${id}`);
  },

  // 创建项目
  create: async (data: CreateProjectData): Promise<Project> => {
    return api.post('/projects', data);
  },

  // 更新项目
  update: async (id: number, data: UpdateProjectData): Promise<Project> => {
    return api.put(`/projects/${id}`, data);
  },

  // 删除项目
  delete: async (id: number): Promise<void> => {
    return api.delete(`/projects/${id}`);
  },
};

export default projectService;

File: backend/api/routes/projects.py
"""
项目管理API路由
"""

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional

from ..database import get_db
from ..models import User, Project
from ..schemas import ProjectCreate, ProjectUpdate, ProjectPublic, ProjectList
from ..utils.dependencies import get_current_active_user

router = APIRouter(prefix="/projects", tags=["Projects"])


@router.post("", response_model=ProjectPublic, status_code=status.HTTP_201_CREATED)
async def create_project(
    data: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """创建新项目"""
    project = Project(
        user_id=current_user.id,
        name=data.name,
        description=data.description,
        config=data.config,
        status="draft",
    )
    db.add(project)
    db.commit()
    db.refresh(project)
    return project


@router.get("", response_model=ProjectList)
async def list_projects(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    status_filter: Optional[str] = Query(None, alias="status"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """获取当前用户的项目列表"""
    query = db.query(Project).filter(Project.user_id == current_user.id)
    if status_filter:
        query = query.filter(Project.status == status_filter)
    total = query.count()
    items = query.order_by(Project.updated_at.desc().nullslast(), Project.created_at.desc()) \
                 .offset(skip).limit(limit).all()
    return ProjectList(total=total, items=items)


@router.get("/{project_id}", response_model=ProjectPublic)
async def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """获取项目详情"""
    project = db.query(Project).filter(
        Project.id == project_id,
        Project.user_id == current_user.id,
    ).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    return project


@router.put("/{project_id}", response_model=ProjectPublic)
async def update_project(
    project_id: int,
    data: ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """更新项目"""
    project = db.query(Project).filter(
        Project.id == project_id,
        Project.user_id == current_user.id,
    ).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")

    allowed_fields = {"name", "description", "config", "status"}
    update_data = data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        if field in allowed_fields:
            setattr(project, field, value)

    db.commit()
    db.refresh(project)
    return project


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """删除项目"""
    project = db.query(Project).filter(
        Project.id == project_id,
        Project.user_id == current_user.id,
    ).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")

    db.delete(project)
    db.commit()
    return None
